### DISCUSSION TITLE
Setting x-data based on x-for template index

### DISCUSSION THREAD

**Discussion by aaronstezycki** [alpinejs/alpine#4147] [1. Help] [answered]
Hi all, any help greatly appreciated here!

I've got an x-for looping through some data, and within the template a modal which then asks for some basic information depending on the current index of the loop. 

Is there any way to dynamically set a unique index which can be used as the identifier for the nested modals/alpine components? Obviously this would allow individual modals to be opened and closed within the x-for.

Apologies if I've missed the solution to this in the docs! 

i.e: 
```
<div x-data="{loop= [...]}" >
   <template x-for="(item, index) in Object.values(loop)" :key="index">
      <div x-data="{ modal_{index}: false }"> <!-- x-data modal container with a unique index based on the loop -->
         <div class="component">
            <button @click="modal_{index} = true">...</button>
         </div>
         <div class="modal" x-show="modal_{index}"> 
             ...
         </div>
      </div>
   </template>
</div>
```

Regards, TIA.

**Comment by SimoTod** [comment, marked as answer]
why do you need different names? The whole point of having a different component for each item is that they can use the same parameter names.

```html
<div x-data="{loop=[...]}" >
   <template x-for="item in loop">
      <div x-data="{ show_modal = false }">
         <div class="component">
            <button @click="show_modal = true">...</button>
         </div>
         <div class="modal" x-show="show_modal"> 
             ...
         </div>
      </div>
   </template>
</div>
```

Not strictly relevant to the question but don't use `:key="index"` as that is the default behaviour so it's redundant. If your array changes over time and needs to keep the DOM state (for example, there are input fields), you need to use a unique property of the item as a key.

If you really need to, Alpine uses the standard javascript syntax. The way you create dynamic property names in javascipt is
```js
{['modal_ ' + index]: false}
```

**Reply by aaronstezycki** [reply]
Hi, okay! will remove the `:key="index"` as I'd seen it in another example and thought it was a useful addition. :)

Where I have it working atm (with effectively this exact example above) whenever I click any of the modal buttons that have been rendered in the loop, it ends up opening *all* of the modals instead of just the modal that affects that current index of the loop. So the unique indexes are just so that the buttons can target the associated modal.

**Reply by aaronstezycki** [reply]
Ah, okay I've seen what I've done in my current working example that causes all modals to open instead of each modal for each loop item. 

It seems it's rather easy to make a silly mistake when mixing lots of JS, accessibility attributes, html and a large amount DOM nodes together.

As a side note, in your experience is it far easier to just abstract everything into dedicated JS files using `Alpine.data(...)` instead of embedding the alpine code directly into component files?

**Reply by ekwoka** [reply]
Generally we extract them. Different frameworks might provide better ways to work with them more inline, but bundling all your data with Alpine.data is the least error prone as things get more complicated. Though smaller ones in the x-data is fine.

**Comment by ekwoka** [comment]
You don't even need unique names here.

But you should look at dynamic property names on MDN for object accessing and object literals.
